core/arch/riscv/kernel/thread.c

```diff
@@ -523,3 +524,231 @@ bool thread_init_stack(uint32_t thread_id, vaddr_t sp)
 	threads[thread_id].stack_va_end = sp;
 	return true;
 }
+
+#ifdef CFG_WITH_PAGER
+static void init_thread_stacks(void)
+{
+	size_t n = 0;
+
+	/*
+	 * Allocate virtual memory for thread stacks.
+	 */
+	for (n = 0; n < CFG_NUM_THREADS; n++) {
+		tee_mm_entry_t *mm = NULL;
+		vaddr_t sp = 0;
+		size_t num_pages = 0;
+		struct fobj *fobj = NULL;
+
+		/* Find vmem for thread stack and its protection gap */
+		mm = tee_mm_alloc(&tee_mm_vcore,
+				  SMALL_PAGE_SIZE + STACK_THREAD_SIZE);
+		assert(mm);
+
+		/* Claim eventual physical page */
+		tee_pager_add_pages(tee_mm_get_smem(mm), tee_mm_get_size(mm),
+				    true);
+
+		num_pages = tee_mm_get_bytes(mm) / SMALL_PAGE_SIZE - 1;
+		fobj = fobj_locked_paged_alloc(num_pages);
+
+		/* Add the region to the pager */
+		tee_pager_add_core_region(tee_mm_get_smem(mm) + SMALL_PAGE_SIZE,
+					  PAGED_REGION_TYPE_LOCK, fobj);
+		fobj_put(fobj);
+
+		/* init effective stack */
+		sp = tee_mm_get_smem(mm) + tee_mm_get_bytes(mm);
+		asan_tag_access((void *)tee_mm_get_smem(mm), (void *)sp);
+		if (!thread_init_stack(n, sp))
+			panic("init stack failed");
+	}
+}
+#else
+static void init_thread_stacks(void)
+{
+	size_t n;
+
+	/* Assign the thread stacks */
+	for (n = 0; n < CFG_NUM_THREADS; n++) {
+		if (!thread_init_stack(n, GET_STACK_BOTTOM(stack_thread, n)))
+			panic("thread_init_stack failed");
+	}
+}
+#endif /*CFG_WITH_PAGER*/
+
+void thread_init_primary(void)
+{
+	/* Initialize canaries around the stacks */
+	init_canaries();
+#if 0
+	init_user_kcode();
+#endif
+}
+
+static void init_sec_mon_stack(size_t pos __maybe_unused)
+{
+}
+
+static void thread_unhandled_trap(struct thread_trap_frame *frame __unused)
+{
+	panic("unhandled exception");
+}
+
+static void dump_trap_frame(struct thread_trap_frame *frame)
+{
+	IMSG(
+	" x0  0x%016" PRIxPTR " ra  0x%016" PRIxPTR
```

Comment:
Please use `lx` instead of `PRIxPTR` since all the registers are defined as `unsigned long`.